tutorials/distributed_data_cache/c++11/home_automation_subscriber.cxx

```diff
@@ -24,97 +24,88 @@ void wait_for_input(std::string query)
+    print_samples(
+            reader.select()
+                    .content(dds::sub::Query(reader, "room_name = 'LivingRoom'"))
+                    .read());
 
 
     wait_for_input(
             "obtain all the samples of the instance with the attribute "
             "sensor_name set to Window1.");
     dds::core::InstanceHandle window1_handle =
             reader.lookup_instance(DeviceStatus("Window1", "LivingRoom", true));
-    samples = reader.select().instance(window1_handle).read();
-    std::for_each(samples.begin(), samples.end(), print_sample);
+    print_samples(reader.select().instance(window1_handle).read());
 
 
     wait_for_input("obtain all the samples that you have not read yet.");
-    samples = reader.select()
-            .state(dds::sub::status::SampleState::not_read())
-            .read();
-    std::for_each(samples.begin(), samples.end(), print_sample);
+    print_samples(
+            reader.select()
+                    .state(dds::sub::status::SampleState::not_read())
+                    .read());
 
 
     wait_for_input("obtain again, all the samples that you have not read yet.");
-    samples = reader.select()
-            .state(dds::sub::status::SampleState::not_read())
-            .read();
-    std::for_each(samples.begin(), samples.end(), print_sample);
+    print_samples(
+            reader.select()
+                .state(dds::sub::status::SampleState::not_read())
+                .read());
 
 
     wait_for_input("obtain the new instances.");
-    samples = reader.select()
-            .state(dds::sub::status::ViewState::new_view())
-            .read();
-    std::for_each(samples.begin(), samples.end(), print_sample);
+    print_samples(
+            reader.select()
+                    .state(dds::sub::status::ViewState::new_view())
+                    .read());
 
 
     wait_for_input(
             "obtain the sensor names of the instances that are not alive.");
-    std::vector<std::string> sensor_names;
-    samples = reader.select()
-            .state(dds::sub::status::InstanceState::not_alive_mask())
-            .read();
-    std::for_each(
-            samples.begin(),
-            samples.end(),
-            [&sensor_names, &reader]
-            (const rti::sub::LoanedSample<DeviceStatus> &sample) {
-                DeviceStatus key_holder;
-                reader.key_value(key_holder, sample.info().instance_handle());
-                sensor_names.push_back(key_holder.sensor_name());
-            });
-    std::for_each(
-            sensor_names.begin(),
-            sensor_names.end(),
-            [](const std::string &sensor_name) {
-                std::cout << sensor_name << std::endl;
-            });
+    DeviceStatus key_holder;
+    dds::sub::LoanedSamples<DeviceStatus> samples =
+            reader.select()
+                    .state(dds::sub::status::InstanceState::not_alive_mask())
+                    .read();
+
+    for (const auto &sample : samples) {
```

Comment:
`const auto&` here is not 100% correct (and can give a warning), because the iterator returns a lightweight temporary object by value (the LoanedSample), not a reference.

Use `auto sample : samples` (`auto&& sample : samples` is the "most" correct and maintainable option because it always binds the variable using the appropriate option (`const&`  or by value, depending on the type of the expression) but to keep the example simpler let's use `auto sample`.)